<a href="https://colab.research.google.com/github/kinjalkothari/socialintelligence/blob/main/01_data_understanding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)


In [4]:
data_path= "LegitReachResearchD2CDataFashionAndApparel.csv"
df = pd.read_csv(data_path)
df.head()


,website,company,country,state,platform,technology_spend,sales_revenue,estimated_visits,website_created_at,emails,telephones,vertical,employees,product_page_count,variant_count,average_product_price,theme,currency,language,appsInstalled,technology,x,twitter,tiktok,facebook,linkedin,pinterest,instagram,youtube,vimeo,whatsapp,yelp,Snapchat,total_affected_elements,total_violation_types,severity_critical,severity_serious,severity_moderate,severity_minor,contrast_errors,aria_errors,accessibility_errors,total_errors
0,intl.nykaafashion.com,Nykaa Fashion Ltd,US,NaN,Shopify,153.970,837258,73117.000,2023-01-27T19:12:27,"support@nykaafashion.com, grievanceofficer@nyk...",NaN,Fashion & Apparel,NaN,25250,132969,200.66,Impulse,usd,English,"ESC Size Charts, qikify Mega Menu & Navigation...","Captcha, Facebook Pixel, Gokwik Kwikpass, Hero...",NaN,https://twitter.com/nykaafashion,NaN,https://facebook.com/nykaafashion,NaN,https://pinterest.com/nykaafashion,https://instagram.com/nykaafashion,NaN,NaN,NaN,NaN,NaN,13.000,4.000,5.000,8.000,0.000,0.000,5.000,0.000,8.000,13.000
1,merch.sixflags.com,Six Flags Entertainment Corporation,US,TX,Shopify,123.170,639915,47933.000,2021-03-03T19:02:06,"copyrightagent@sixflags.com, sixflags@wearsthe...","1-210-697-5050,+1-301-249-1500,+1-972-595-5000...",Fashion & Apparel,NaN,567,1617,31.23,Origin,usd,English,"Customily Product Personalizer, Printful Size ...","Amazon Fba, Captcha, Gokwik Kwikpass, Printful",NaN,https://twitter.com/sixflags,NaN,https://facebook.com/sixflags,https://linkedin.com/company/six-flags,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.000,2.000,5.000,1.000,0.000,0.000,5.000,0.000,1.000,6.000
2,usastore.thebeatles.com,Apple Corps Ltd,US,NaN,Shopify,381.970,1061528,90631.000,2020-08-30T13:29:12,"privacy.officer@umusic.com, thebeatles@umgstor...","1-207-761-9600,+44-207-7619600",Fashion & Apparel,NaN,1058,2709,67.6,Rhythm,usd,English,"ESC Size Charts, Printful Size Guide,Size Char...","Attentive, Captcha, Facebook Pixel, Gokwik Kwi...",NaN,https://twitter.com/thebeatles,NaN,https://facebook.com/beatles,NaN,NaN,https://instagram.com/thebeatles,https://www.youtube.com/channel/UCc4K7bAqpdBP8...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,shop.atlantamagazine.com,Atlanta Magazine,US,GA,Shopify,133.480,5118,0.000,2023-11-30T11:06:47,"mbikoff@atlantamagazine.com, trehagen@atlantam...","1-404-527-5575,+1-800-351-7221,+1-404-843-9838...",Fashion & Apparel,NaN,714,1482,54.49,Impact,usd,English,"Printful Size Guide,Size Chart, Printful: Prin...","Amazon Fba, Captcha, Gokwik Kwikpass, Klaviyo,...",https://x.com/atlantamagazine,https://twitter.com/atlantamagazine,NaN,https://facebook.com/atlantamag,https://linkedin.com/company/atlantamagazine,NaN,https://instagram.com/atlantamagazine,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,shop.thelibertydaily.com,The Liberty Daily,US,PA,Shopify,214.900,90758,5502.000,2021-03-19T21:02:51,"paul@theprb.co, thelibertydaily@gmail.com",NaN,Fashion & Apparel,NaN,264,4632,25.42,Dawn,usd,English,"GLO Color Swatch & Bundles, Vario ‑ Variation ...","Captcha, Gokwik Kwikpass, Klaviyo",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [5]:
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

df.info()


Rows: 14288
Columns: 43
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14288 entries, 0 to 14287
Data columns (total 43 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   website                  14288 non-null  object 
 1   company                  12888 non-null  object 
 2   country                  14288 non-null  object 
 3   state                    3513 non-null   object 
 4   platform                 14288 non-null  object 
 5   technology_spend         14288 non-null  float64
 6   sales_revenue            14288 non-null  int64  
 7   estimated_visits         14288 non-null  float64
 8   website_created_at       14288 non-null  object 
 9   emails                   13631 non-null  object 
 10  telephones               3270 non-null   object 
 11  vertical                 14288 non-null  object 
 12  employees                33 non-null     float64
 13  product_page_count       14288 non-null  object 
 14

The dataset contains X rows and Y columns, representing brand-level digital presence and business performance indicators. Column types include numerical metrics and categorical descriptors


In [6]:
numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()

print("Numeric columns:", numeric_cols)
print("Categorical columns:", categorical_cols)


Numeric columns: ['technology_spend', 'sales_revenue', 'estimated_visits', 'employees', 'whatsapp', 'yelp', 'Snapchat', 'total_affected_elements', 'total_violation_types', 'severity_critical', 'severity_serious', 'severity_moderate', 'severity_minor', 'contrast_errors', 'aria_errors', 'accessibility_errors', 'total_errors']
Categorical columns: ['website', 'company', 'country', 'state', 'platform', 'website_created_at', 'emails', 'telephones', 'vertical', 'product_page_count', 'variant_count', 'average_product_price', 'theme', 'currency', 'language', 'appsInstalled', 'technology', 'x', 'twitter', 'tiktok', 'facebook', 'linkedin', 'pinterest', 'instagram', 'youtube', 'vimeo']


Numerical columns primarily represent performance metrics and engineered indicators, while categorical columns describe brand attributes and platform identifiers.


In [7]:
missing_df = (
    df.isnull()
    .mean()
    .sort_values(ascending=False)
    .reset_index()
)

missing_df.columns = ["column", "missing_ratio"]
missing_df


,column,missing_ratio
0,whatsapp,1.000
1,Snapchat,1.000
2,yelp,1.000
3,vimeo,1.000
4,employees,0.998
5,x,0.992
6,linkedin,0.956
7,accessibility_errors,0.924
8,total_violation_types,0.924
9,severity_serious,0.924


Social media platform columns show missing values that logically indicate absence or inactivity, whereas missing business metrics require careful treatment to avoid bias.

In [9]:
cols_to_drop = ['whatsapp', 'Snapchat', 'yelp', 'vimeo']
df = df.drop(columns=cols_to_drop)
print(f"Dropped columns: {cols_to_drop}")
print(f"New DataFrame shape: {df.shape}")

Dropped columns: ['whatsapp', 'Snapchat', 'yelp', 'vimeo']
New DataFrame shape: (14288, 39)


These columns were dropped as their missing percentage in dataset is 100%.

In [10]:
social_keywords = [
    "instagram", "facebook", "twitter", "linkedin",
    "tiktok", "youtube", "pinterest"
]

social_cols = [
    col for col in df.columns
    if any(keyword in col.lower() for keyword in social_keywords)
]

social_cols


['twitter',
 'tiktok',
 'facebook',
 'linkedin',
 'pinterest',
 'instagram',
 'youtube']

These columns capture brand presence across multiple social platforms. Missing values will later be treated as inactive presence (0) rather than true nulls.

In [18]:
TARGET = "estimated_visits"

df[TARGET].describe()


,estimated_visits
count,14288.000
mean,6703.685
std,30185.882
min,0.000
25%,734.000
50%,1394.000
75%,3629.000
max,1835619.000


estimated_visits is selected as the target variable because it serves as a proxy for brand visibility and customer acquisition. The distribution shows moderate skewness, which will be addressed during modeling.


**Key Observations**
1. Dataset Structure and Scale

The dataset comprises 14,288 observations and 43 initial features, representing brand-level attributes, digital presence indicators, and performance metrics. After removing columns with no informational value, the effective feature set is reduced to 39 columns, maintaining full row coverage.

2. Data Types and Structural Inconsistencies

An initial inspection of data types reveals a mixture of object, float64, and int64 features. Notably, variables such as product_page_count, variant_count, and average_product_price are stored as object types despite representing numerical quantities. This mismatch indicates the need for explicit type coercion and validation prior to modeling, as untreated object types would prevent direct numerical analysis and model ingestion.

3. Missing Data Characteristics

Four social platform columns—WhatsApp, Snapchat, Yelp, and Vimeo—exhibit 100% missing values and were therefore removed due to the absence of any informational content.

Beyond these, a substantial number of features display high missingness, particularly:

Social media presence indicators (e.g., X, LinkedIn, YouTube, Pinterest, Twitter, TikTok)

Contact and organizational metadata (e.g., telephones, emails, employees, state)

Accessibility-related metrics (e.g., total_affected_elements, severity_critical)

Missingness across these variables ranges from moderate levels (~4–10%) to extreme cases exceeding 90–99%, suggesting that:

Certain attributes are selectively available rather than universally collected.

NaN values in social platform columns likely encode absence or inactivity, rather than data corruption.

4. Target Variable: estimated_visits

The variable estimated_visits has been selected as the primary modeling target, serving as a proxy for brand visibility and user acquisition.

Key properties:

No missing values (14,288 non-null observations).

Highly right-skewed distribution, with the mean (~6,703) substantially exceeding the median (~1,394).

Large variance, with values ranging from 0 to over 1.8 million, indicating significant heterogeneity in website traffic across brands.

These characteristics suggest the presence of outliers and non-normality, which will influence downstream modeling choices.

Risks and Modeling Considerations
1. Data Type Conversion Risk

Incorrect data types for numerically meaningful columns may lead to:

Model incompatibility

Silent exclusion of features

Incorrect statistical summaries

Careful parsing and coercion will be required to ensure valid numerical representations without unintentionally discarding meaningful observations.

2. High Missingness and Information Loss

The pervasive nature of missing data introduces several challenges:

Row-wise deletion is infeasible, as it would remove a majority of observations.

Imputation for extremely sparse features (e.g., employees, x) risks introducing artificial patterns and bias.

Feature selection and missingness-aware handling will be critical to preserve signal quality.

3. Interpretation of Social Media Absence

For social platform indicators, missing values are interpreted as absence or inactivity rather than unknown values. The chosen encoding strategy must preserve this semantic meaning while avoiding spurious correlations introduced by naive imputation.

4. Bias in Accessibility Metrics

Accessibility-related variables exhibit very high missing rates, suggesting they may only be measured for a limited subset of websites. Direct inclusion of these features could bias model behavior toward this subset unless handled with care.

5. Target Skewness

The strong right skew in estimated_visits violates assumptions of many regression models. Without appropriate transformation (e.g., logarithmic scaling), models may:

Overemphasize extreme values

Produce biased error estimates

Exhibit reduced generalization performance

6. High Cardinality in Categorical Features

Several categorical variables (e.g., website, company, technology, appsInstalled, emails) are likely to have high cardinality. Naive one-hot encoding would dramatically increase dimensionality, leading to sparsity and overfitting. Alternative encoding strategies will be required in subsequent preprocessing stages.

In [20]:
df.to_csv("../data_after_understanding.csv", index=False)
